cpp/include/cudf/scalar/scalar_device_view.cuh

```diff
@@ -0,0 +1,187 @@
+                            ///< boolean representing validity of the value.
+
+  scalar_device_view_base(data_type type, bool* is_valid)
+    : _type(type), _is_valid(is_valid) 
+  {}
+};
+
+template <typename T>
+class alignas(16) primitive_scalar_device_view
+    : public detail::scalar_device_view_base {
+ public:
+
+  /**---------------------------------------------------------------------------*
+   * @brief Returns reference to element at the specified index.
+   *---------------------------------------------------------------------------**/
+  __device__ T& value() noexcept {
+    return *get();
+  }
+
+  __device__ T const& value() const noexcept {
+    return *get();
+  }
+
+  __device__ void set_value(T value) {
+    *_data = value;
+  }
+
+  __device__ T* get() noexcept {
+    return static_cast<T*>(_data);
+  }
+
+  __device__ T const* get() const noexcept {
+    return static_cast<T const*>(_data);
+  }
+
+ protected:
+  T * _data{};      ///< Pointer to device memory containing the value
+
+  primitive_scalar_device_view(data_type type, T* data, bool* is_valid)
+   : detail::scalar_device_view_base(type, is_valid)
+   , _data(data)
+  {}
+};
+
+}  // namespace detail
+
+
+template <typename T>
+class alignas(16) numeric_scalar_device_view
+    : public detail::primitive_scalar_device_view<T>
+{
+ public:
+  using ValueType = T;
+
+  numeric_scalar_device_view(data_type type, T* data, bool* is_valid) 
+    : detail::primitive_scalar_device_view<T>(type, data, is_valid)
+  {}
+};
+
+class alignas(16) string_scalar_device_view
+    : public detail::scalar_device_view_base {
+ public:
+  using ValueType = cudf::string_view;
+
+  string_scalar_device_view(data_type type, const char* data, bool* is_valid,
+                            size_type size) 
+    : detail::scalar_device_view_base(type, is_valid), _data(data), _size(size)
+  {}
+
+  /**---------------------------------------------------------------------------*
+   * @brief Returns string_view of the value of this scalar.
+   *---------------------------------------------------------------------------**/
+  __device__ string_view value() const noexcept {
+    return string_view(this->get(), _size);
+  }
+
+  __device__ char const* get() const noexcept {
+    return static_cast<char const*>(_data);
+  }
+
+ private:
+  const char* _data{};  ///< Pointer to device memory containing the value
+  size_type _size;      ///< Length of the string
+};
+
+template <typename T>
+class alignas(16) timestamp_scalar_device_view
+    : public detail::primitive_scalar_device_view<T>
+{
+ public:
+  using ValueType = T;
+
+  timestamp_scalar_device_view(data_type type, T* data, bool* is_valid) 
+    : detail::primitive_scalar_device_view<T>(type, data, is_valid)
+  {}
+};
+
+template <typename T>
+auto get_scalar_device_view(numeric_scalar<T>& s) {
+  return numeric_scalar_device_view<T>(s.type(), s.data(), s.valid_mask());
+}
+
+auto get_scalar_device_view(string_scalar& s) {
```

Comment:
Since these don't need to allocate device memory (unlike with `column/table_device_view`), I think it makes sense for these to be conversion operators instead of needing free functions. E.g.,
```
class numeric_scalar{
   operator numeric_scalar_device_view(){
   ...
   }
};
```

Reply:
The reason I made this a separate function instead of conversion operator is so that type dispatch functors need not specify the type of scalar they want to get.
With free functions:
```c++
template <typename scalar_type>
__global__ kernel(scalar_type s) {...}

template <typename T, typename scalar_type = scalar_type_t<T>>
void operator()(scalar_type& s) {
  kernel<<<>>>(get_scalar_device_view(s));
}
```
Using implicit conversion member function, it wouldn't be possible to instantiate the kernel template. We could, however, explicitly instantiate and call the member function `device_view()`
```c++
  kernel<<<>>>(s.device_view());
```
But implementing this proved a bit difficult because these need to be templated so they have to be in the header; otherwise there would be code duplication trying to instantiate the method for each type in the source. So `scalar_device_view.cuh` would need to be included in `scalar.hpp`, which I'm trying to avoid because then everything that includes `scalar.hpp` would need to be `.cu`

Reply:
> Using implicit conversion member function, it wouldn't be possible to instantiate the kernel template. 

Good point.

Though, I wouldn't expect something like this to be possible:
```
template <typename T, typename scalar_type = scalar_type_t<T>>
void operator()(scalar_type& s) {
  kernel<<<>>>(get_scalar_device_view(s));
}
```

Instead, I would expect something like this:
```
template <typename T, typename scalar_type = scalar_type_t<T>>
void operator()(scalar& s) {
  d = static_cast<scalar_type&>(s);
  kernel<<<>>>(get_scalar_device_view(d));
}
```
i.e., a `cudf::scalar` base class reference would be passed into the type dispatched function, and then it would need to get casted to the appropriate derived type.